extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
-            Set<MethodInfo> blockingMethods = gatherBlockingMethods(service);
-            for (MethodInfo method : blockingMethods) {
-                item.registerBlockingMethod(method.name());
+            Set<String> blockingMethods = gatherBlockingMethods(service, index);
+            for (String method : blockingMethods) {
+                item.registerBlockingMethod(method);
             }
             bindables.produce(item);
         }
     }
 
-    private Set<MethodInfo> gatherBlockingMethods(ClassInfo service) {
-        Set<MethodInfo> result = new HashSet<>();
-        boolean blockingClass = service.classAnnotation(GrpcDotNames.BLOCKING) != null;
-        if (blockingClass && service.classAnnotation(GrpcDotNames.NON_BLOCKING) != null) {
+    /**
+     * Generate list of {@link ClassInfo} with {@code service} as the first element and the class implementing
+     * {@code io.grpc.BindableService} (the protobuf generated {@code *ImplBase}) as the last one.
+     */
+    private static List<ClassInfo> classHierarchy(ClassInfo service, IndexView index) {
+        List<ClassInfo> collected = new ArrayList<>();
+        for (ClassInfo ci = service; ci != null;) {
+            collected.add(ci);
+
+            // Stop at the class that implements io.grpc.BindableService or
+            if (ci.interfaceNames().contains(GrpcDotNames.BINDABLE_SERVICE)) {
+                break;
+            }
+
+            DotName superName = ci.superName();
+            if (superName == null) {
+                break;
+            }
+            ci = index.getClassByName(superName);
+        }
+        return collected;
+    }
+
+    private static Set<DotName> relevantAnnotations(List<ClassInfo> classHierarchy) {
+        Set<DotName> relevant = new HashSet<>();
+        relevant.add(GrpcDotNames.BLOCKING);
+        relevant.add(GrpcDotNames.NON_BLOCKING);
+        relevant.add(TRANSACTIONAL);
+
+        return classHierarchy.stream()
+                .flatMap(ci -> ci.classAnnotations().stream())
+                .map(AnnotationInstance::name)
+                .filter(relevant::contains)
+                .collect(Collectors.toSet());
+    }
+
+    private static Set<DotName> relevantAnnotations(List<ClassInfo> classHierarchy, MethodInfo method) {
+        Set<DotName> relevant = new HashSet<>();
+        relevant.add(GrpcDotNames.BLOCKING);
+        relevant.add(GrpcDotNames.NON_BLOCKING);
+        relevant.add(TRANSACTIONAL);
+
+        return classHierarchy.stream()
+                .map(ci -> ci.method(method.name(), method.parameters().toArray(new Type[0])))
+                .filter(Objects::nonNull)
+                .flatMap(mi -> mi.annotations().stream())
+                .map(AnnotationInstance::name)
+                .filter(relevant::contains)
+                .collect(Collectors.toSet());
+    }
+
+    private Set<String> gatherBlockingMethods(ClassInfo service, IndexView index) {
```

Comment:
what if a class has two methods with the same name?

Reply:
hmm... I see we are registering blocking methods by name currently too

Reply:
Yea - duplicate names aren't allowed in protobuf. I thought, it's easier to read with that change to `String`.